testing/integration/upgrade_test.go

```diff
@@ -273,7 +274,165 @@ func TestStandaloneUpgrade(t *testing.T) {
+	t.Logf("Agent installation output: %q", string(output))
+	require.NoError(t, err)
+
+	c := agentFixture.Client()
+
+	require.Eventually(t, func() bool {
+		err := c.Connect(ctx)
+		if err != nil {
+			t.Logf("connecting client to agent: %v", err)
+			return false
+		}
+		defer c.Disconnect()
+		state, err := c.State(ctx)
+		if err != nil {
+			t.Logf("error getting the agent state: %v", err)
+			return false
+		}
+		t.Logf("agent state: %+v", state)
+		return state.State == cproto.State_HEALTHY
+	}, 2*time.Minute, 10*time.Second, "Agent never became healthy")
+
+	aac := tools.NewArtifactAPIClient()
+	vList, err := aac.GetVersions(ctx)
+	require.NoError(t, err, "error retrieving versions from Artifact API")
+	require.NotNil(t, vList)
+
+	sortedParsedVersions := make(version.SortableParsedVersions, 0, len(vList.Versions))
+	for _, v := range vList.Versions {
+		pv, err := version.ParseVersion(v)
+		require.NoErrorf(t, err, "invalid version retrieved from artifact API: %q", v)
+		sortedParsedVersions = append(sortedParsedVersions, pv)
+	}
+
+	require.NotEmpty(t, sortedParsedVersions)
+
+	// normally the output of the versions returned by artifact API is already sorted in ascending order,
+	// if we want to sort in descending order we could use
+	sort.Sort(sort.Reverse(sortedParsedVersions))
+
+	var latestSnapshotVersion *version.ParsedSemVer
+	// fetch the latest SNAPSHOT build
+	for _, pv := range sortedParsedVersions {
+		if pv.IsSnapshot() {
+			latestSnapshotVersion = pv
+			break
+		}
+	}
+
+	require.NotNil(t, latestSnapshotVersion)
+
+	// get all the builds of the snapshot version (need to pass x.y.z-SNAPSHOT format)
+	builds, err := aac.GetBuildsForVersion(ctx, latestSnapshotVersion.VersionWithPrerelease())
+	require.NoError(t, err)
+	// TODO if we don't have at least 2 builds, select the next older snapshot build
+	require.Greater(t, len(builds.Builds), 1)
+
+	// take the penultimate build of the snapshot (the builds are ordered from most to least recent)
+	upgradeVersionString := builds.Builds[1]
+
+	t.Logf("Targeting build %q of version %q", upgradeVersionString, latestSnapshotVersion)
+
+	buildDetails, err := aac.GetBuildDetails(ctx, latestSnapshotVersion.VersionWithPrerelease(), upgradeVersionString)
+	require.NoErrorf(t, err, "error accessing build details for version %q and buildID %q", latestSnapshotVersion.Original(), upgradeVersionString)
+	require.NotNil(t, buildDetails)
+	agentProject, ok := buildDetails.Build.Projects["elastic-agent"]
+	require.Truef(t, ok, "elastic agent project not found in version %q build %q", latestSnapshotVersion.Original(), upgradeVersionString)
+	t.Logf("agent build details: %+v", agentProject)
+	t.Logf("expected agent commit hash: %q", agentProject.CommitHash)
+	expectedAgentHashAfterUpgrade := agentProject.CommitHash
+
+	// Workaround until issue with Artifact API build commit hash are resolved
+	actualAgentHashAfterUpgrade := extractCommitHashFromArtifact(t, ctx, latestSnapshotVersion, agentProject)
```

Comment:
Is there a specific reason why you want to install the penultimate SNAPSHOT build? This is from the upgrade to specific snapshot integration test but you could simply use the latest snapshot or the latest released version, right ?

Reply:
i just took normal test as a base and added custom flags for my usecase, i havent changed a thing so it's better to ask probably @ycombinator about why this was chosen